### PR TITLE
add pagination and base_url to JSON query_runner

### PR DESCRIPTION
**Pull request by Ken-Michalak**
## What type of PR is this? 

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] New Query Runner (Data Source) 
- [ ] New Alert Destination
- [ ] Other

## Description

Add the ability to paginate the JSON Query Runner and configure a base URL.
Support [Google's pagination design pattern](https://cloud.google.com/apis/design/design_patterns#list_pagination), with `pagination: [next_page_token, page_token]`.
Support a next URL, which is sometimes used in HATEOAS, with `pagination: _links.next.href`.

## How is this tested?

- [x] Unit tests (pytest, jest)
- [ ] E2E Tests (Cypress)
- [ ] Manually
- [ ] N/A

## Related Tickets & Documents
Original PR: https://github.com/getredash/redash/pull/5067

### CODE REVIEW

**Comment by Ken-Michalak** [comment]
@justinclift Here's the new pr.

I'm not sure if this is the best way to configure it, especially if someone wants to support other patterns later, like page numbers or offsets without anything in response for what to pass next. I was wondering if there were other ideas there or if there's something like that on another data source.

**Comment by justinclift** [comment]
@guidopetri @eradman Any interest in reviewing this? :smile:

**Comment by justinclift** [comment]
Ahhh, looks like this has a formatting issue that needs fixing:

```
Run ruff check .
redash/query_runner/json_ds.py:1:1: I001 [*] Import block is un-sorted or un-formatted
```

**Comment by guidopetri** [comment]
Could we also add a test to make it clearer what is going on, and what the expected behavior is?

**Comment by Ken-Michalak** [comment]
@guidopetri I restructured the configuration, which hopefully makes a bit more sense.

For a config like this:
```yaml
pagination:
  type: url
  #path: _links.next.href
```
As long as the API response has something in `_links.next.href`, it will follow it, do another request, and include those results.

For the other pattern:
```yaml
pagination:
    type: token
    #fields: [next_page_token, page_token]
```
It does the same kind of thing, where as long as `next_page_token` isn't `""` or `null`, it continues getting the next page of results by putting that token in the `page_token` parameter.

I looked into adding a test, but it doesn't look like there are any tests for the json data source at all yet, and this would be pretty complicated to mock. I was just testing it manually by querying a [Chargebee API](https://apidocs.chargebee.com/docs/api/list-ops?lang=curl) and a [Vimeo VHX API](https://dev.vhx.tv/docs/api/#product-list).

File: redash/query_runner/json_ds.py
Comment (guidopetri) [inline redash/query_runner/json_ds.py:207]: Having this many nested `if` statements is kind of a code smell. Maybe we should split this up into multiple functions?

File: redash/query_runner/json_ds.py
Comment (guidopetri) [inline redash/query_runner/json_ds.py:210]: I've never seen either of these two pagination patterns. I don't think there's a "standard" either, is there? I imagine every API has its own set of pagination rules that it follows. If that's the case, I'm not sure it makes sense to write in pagination rules, and rather we should emphasize to the end users that querying an API for JSONs directly like this is meant for testing and not large-scale analysis.

Especially since if too much data is returned, it's possible that the worker goes OOM. 😅

File: redash/query_runner/json_ds.py
Comment (guidopetri) [inline redash/query_runner/json_ds.py:76]: Won't this return on the first instance of `current_path not in response`? I am not sure having the return within the `while` loop is what we want, but rather we should rewrite this so that there's only one return statement in the function akin to `return response or default` , or something similar.

**Review by guidopetri — CHANGES_REQUESTED** [review]
I very much don't feel comfortable merging this in without a test. To be honest I'm not sold on the idea of implementing pagination here in the first place. As mentioned in the comments below, I'm not sure there's a standard for pagination, and if we were to implement every pagination ruleset out there we'd have a very large codebase 😅 but if we do want to merge this in I would very kindly request that we somehow add tests, even if we have to write some from scratch for the JSON query runner to start out with.

File: redash/query_runner/json_ds.py
Comment (Ken-Michalak) [reply redash/query_runner/json_ds.py:210]: Some variant of these 2 seem to be the most common, because they don't require specifics of implementation, like having to add to the page number or offset yourself. (The `type` could be named something different if you have a better idea though.) That's why Google outlines the next_page_token/page_token as a best practice. It's just that sometimes the names are different, like Chargebee uses an offset parameter, but rather than having to calculate it, it returns the next offset in the response. Google Analytics, AWS, and K8s also follow that pattern.

Shopify, SalesForce, Facebook's Graph API, and Trino all pass a full URL for the next page in some field of the response, which would be covered by the `url` type. HATEOAS is a standard and does define that the next page would be a link somewhere in the response. Although, at some point, it might be good to use a library instead of `_apply_path_search` to handle more complex cases.

Other types could be added later for `pageNumber` or something, but I don't really have a need for them right now, and that is where it could be pretty difficult to cover different APIs.

I was wondering about the memory thing, but it doesn't look like there's any other way that large result sets are handled elsewhere. The Trino data source does a `fetchall` and then processes all of the results at once. This at least does the `_normalize_json` per page, which could trim things down.

**Comment by Ken-Michalak** [comment]
Odd, I didn't see any linting errors, but yeah, black still changed some things.

File: redash/query_runner/json_ds.py
Comment (guidopetri) [reply redash/query_runner/json_ds.py:214]: In my opinion, yes - definitely worth it if it makes the code easier to follow.

File: redash/query_runner/json_ds.py
Comment (guidopetri) [reply redash/query_runner/json_ds.py]: Fair points; thanks for elaborating.

File: tests/query_runner/test_json_ds.py
Comment (guidopetri) [inline tests/query_runner/test_json_ds.py:53]: Ah, sorry, I should have specified. We've been aiming for pytest tests instead of unittest; if it's not too much trouble, would you mind converting to those? I won't hold it against you if you're not feeling up to it though. I appreciate the work you've put into this PR and already recognize you've gone above and beyond :)

**Review by guidopetri — APPROVED** [review]
Approving, but we'll need the PR to be updated with the latest `master` to merge. I'll enable automerge so whenever you push a merge commit updating the branch, it should merge automatically once the tests pass.

Thanks again for all your work @Ken-Michalak !

File: tests/query_runner/test_json_ds.py
Comment (Ken-Michalak) [reply tests/query_runner/test_json_ds.py:53]: I just used the other tests as an example. It could always be refactored later. It seems like some of that should be added to the contributing guide, like the testing, ruff, and rebasing process. Like are you looking for a single commit per pr, or just going for a linear history?

Comment (guidopetri) [reply tests/query_runner/test_json_ds.py:53]: That's a great point. I don't think we have a good handle on what our contribution guidelines are; I'll try and add some documentation for that to the wiki soon.

**Review by guidopetri — APPROVED** [review]

**Comment by justinclift** [comment]
@Ken-Michalak We have an option turned on for this repo, which requires PR's to be uo-to-date with the master branch before they can be merged.

Normally PRs have a button to "Update to the master branch" (or similar wording), that maintainers can click to automate that.

It seems to be missing from this PR though.  My guess is the PR was created without the "Allow maintainers to update this PR?" setting enabled.

So... I think that means you'll need to update this PR yourself.  That might let it all just get nicely merged automatically after the tests pass. :smile:

**Comment by guidopetri** [comment]
I bypassed it - we didn't touch the poetry files so it should be ok.

Thanks for your contribution @Ken-Michalak !

**Comment by Ken-Michalak** [comment]
> @Ken-Michalak We have an option turned on for this repo, which requires PR's to be uo-to-date with the master branch before they can be merged.
> 
> Normally PRs have a button to "Update to the master branch" (or similar wording), that maintainers can click to automate that.
> 
> It seems to be missing from this PR though. My guess is the PR was created without the "Allow maintainers to update this PR?" setting enabled.
> 
> So... I think that means you'll need to update this PR yourself. That might let it all just get nicely merged automatically after the tests pass. 😄

@justinclift, I actually did that in that last force-push yesterday, but it looks like before @guidopetri approved it again, there were a few other commits added. (Bump @babel/traverse from 7.22.8 to 7.23.2) 🤷🏻 Thanks though.

**Comment by justinclift** [comment]
No worries, it all worked out.  Thanks heaps for getting this done @Ken-Michalak, it seems like this will be useful for people. :smile:

**Pull request merged**
